src/Interpreters/TableOverrideUtils.cpp

```diff
@@ -0,0 +1,169 @@
+#include <Interpreters/TableOverrideUtils.h>
+
+#include <Common/quoteString.h>
+#include <DataTypes/IDataType.h>
+#include <DataTypes/DataTypeFactory.h>
+#include <IO/WriteBufferFromString.h>
+#include <Interpreters/InDepthNodeVisitor.h>
+#include <Interpreters/RequiredSourceColumnsVisitor.h>
+#include <Parsers/ASTColumnDeclaration.h>
+#include <Parsers/ASTCreateQuery.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTTableOverrides.h>
+#include <Parsers/IAST.h>
+#include <Storages/ColumnDefault.h>
+#include <Storages/StorageInMemoryMetadata.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int INVALID_TABLE_OVERRIDE;
+}
+
+namespace
+{
+
+class MaybeNullableColumnsMatcher
+{
+public:
+    using Visitor = ConstInDepthNodeVisitor<MaybeNullableColumnsMatcher, false>;
+    using Data = RequiredSourceColumnsData;
+
+    static bool needChildVisit(const ASTPtr & node, const ASTPtr & child)
+    {
+        if (const auto * f = node->as<ASTFunction>(); f && f->name == "assumeNotNull")
+            return false;
```

Comment:
why do we skip such nodes?

Reply:
This visitor is just a special-casing of the required columns visitor to find nullable columns, this allows users to exclude columns from that if they "know" that it never contains NULL values. For example, if `foo` is a nullable column in the source database, but you want to use it for SAMPLE BY, you do `SAMPLE BY assumeNotNull(foo)`. Just `SAMPLE BY foo` would throw an exception when the column is nullable.